File: services/balance_history_service.py
"""Unified jopacoin balance-history series for the /profile Economy tab chart."""

from __future__ import annotations

import json
from dataclasses import dataclass, field

from config import JOPACOIN_PER_GAME, JOPACOIN_WIN_REWARD

# Source identifiers — must stay in sync with ``utils/drawing/balance_history.py``.
SOURCE_BETS = "bets"
SOURCE_PREDICTIONS = "predictions"
SOURCE_WHEEL = "wheel"
SOURCE_DOUBLE_OR_NOTHING = "double_or_nothing"
SOURCE_TIPS = "tips"
SOURCE_DISBURSE = "disburse"
SOURCE_BONUS = "bonus"
SOURCE_DIG = "dig"


@dataclass
class _Event:
    time: int
    delta: int
    source: str
    detail: dict = field(default_factory=dict)


class BalanceHistoryService:
    """Merge all persisted balance-impacting events for a player into one series.

    Eight sources are chartable today:
      - ``bets``              settled bets (profit)
      - ``predictions``       resolved prediction bets (payout − staked)
      - ``wheel``             Wheel of Fortune spins (result)
      - ``double_or_nothing`` Double or Nothing spins (balance_after − (balance_before + cost))
      - ``tips``              sent (-(amount+fee)) and received (+amount)
      - ``disburse``          nonprofit disbursements received (+amount)
      - ``bonus``             per-match participation + win bonus reconstruction
      - ``dig``               dig actions: earnings, event JC, boss fight wins/losses,
                              abandon refunds, retreat losses, sabotage stake/trap-steal,
                              help rewards (parsed from the ``dig_actions.detail`` JSON)

    Un-persisted sources (shop, admin /givecoin, bankruptcy penalty, garnishment,
    streaming / first-game / exclusion / bomb-pot bonuses, cancelled-match refunds,
    dig cave-in losses, per-dig paid costs) are silently omitted — the series starts
    at 0 and does not pretend to equal live balance.
    """

    def __init__(
        self,
        bet_repo,
        match_repo,
        player_repo,
        prediction_repo,
        disburse_repo,
        tip_repo,
        dig_repo=None,
    ):
        self.bet_repo = bet_repo
        self.match_repo = match_repo
        self.player_repo = player_repo
        self.prediction_repo = prediction_repo
        self.disburse_repo = disburse_repo
        self.tip_repo = tip_repo
        self.dig_repo = dig_repo

    def get_balance_event_series(
        self, discord_id: int, guild_id: int | None = None
    ) -> tuple[list[tuple[int, int, dict]], dict[str, int]]:
        """
        Return ``(series, per_source_totals)``.

        ``series``: list of ``(event_number, cumulative_delta, event_info)`` starting
        at ``event_number=1`` and ``cumulative_delta`` starting at 0. Sorted by time.

        ``event_info``: ``{"time": int, "delta": int, "source": str, "detail": dict}``.

        ``per_source_totals``: ``{source: net_delta}`` for sources with a non-zero
        total. Matches are empty if the player has no recorded activity anywhere.
        """
        events: list[_Event] = []
        events.extend(self._bet_events(discord_id, guild_id))
        events.extend(self._prediction_events(discord_id, guild_id))
        events.extend(self._wheel_events(discord_id, guild_id))
        events.extend(self._double_or_nothing_events(discord_id, guild_id))
        events.extend(self._tip_events(discord_id, guild_id))
        events.extend(self._disburse_events(discord_id, guild_id))
        events.extend(self._bonus_events(discord_id, guild_id))
        events.extend(self._dig_events(discord_id, guild_id))

        if not events:
            return [], {}

        events.sort(key=lambda e: e.time)

        series: list[tuple[int, int, dict]] = []
        totals: dict[str, int] = {}
        cumulative = 0
        for idx, ev in enumerate(events, start=1):
            cumulative += ev.delta
            series.append(
                (
                    idx,
                    cumulative,
                    {
                        "time": ev.time,
                        "delta": ev.delta,
                        "source": ev.source,
                        "detail": ev.detail,
                    },
                )
            )
            totals[ev.source] = totals.get(ev.source, 0) + ev.delta

        per_source_totals = {src: total for src, total in totals.items() if total != 0}
        return series, per_source_totals

    # ── Per-source collectors ────────────────────────────────────────────────

    def _bet_events(self, discord_id: int, guild_id: int | None) -> list[_Event]:
        rows = self.bet_repo.get_player_bet_history(discord_id, guild_id)
        return [
            _Event(
                time=int(row["bet_time"]),
                delta=int(row["profit"]),
                source=SOURCE_BETS,
                detail={
                    "outcome": row["outcome"],
                    "amount": row["amount"],
                    "leverage": row["leverage"],
                    "match_id": row["match_id"],
                },
            )
            for row in rows
        ]

    def _prediction_events(
        self, discord_id: int, guild_id: int | None
    ) -> list[_Event]:
        rows = self.prediction_repo.get_player_prediction_history(discord_id, guild_id)
        out: list[_Event] = []
        for row in rows:
            staked = int(row["total_amount"] or 0)
            payout = int(row["payout"] or 0)
            delta = payout - staked
            # Drop perfect net-zero events so they don't clutter the chart
            # (cancelled predictions with full refund fall out this way).
            if delta == 0:
                continue
            out.append(
                _Event(
                    time=int(row["settle_time"] or 0),
                    delta=delta,
                    source=SOURCE_PREDICTIONS,
                    detail={
                        "outcome": "won" if delta > 0 else "lost",
                        "position": row["position"],
                        "status": row["status"],
                        "prediction_id": row["prediction_id"],
                    },
                )
            )
        return out

    def _wheel_events(self, discord_id: int, guild_id: int | None) -> list[_Event]:
        rows = self.player_repo.get_wheel_spin_history(discord_id, guild_id)
        out: list[_Event] = []
        for row in rows:
            result = int(row["result"])
            if result == 0:
                continue  # "lose a turn" — no balance change, skip
            out.append(
                _Event(
                    time=int(row["spin_time"]),
                    delta=result,
                    source=SOURCE_WHEEL,
                    detail={"outcome": "won" if result > 0 else "lost"},
                )
            )
        return out

    def _double_or_nothing_events(
        self, discord_id: int, guild_id: int | None
    ) -> list[_Event]:
        # The repo method's annotation says ``guild_id: int`` but it routes through
        # ``normalize_guild_id`` which accepts None, matching every other repo here.
        rows = self.player_repo.get_double_or_nothing_history(discord_id, guild_id)
        out: list[_Event] = []
        for row in rows:
            balance_before = int(row["balance_before"])
            balance_after = int(row["balance_after"])
            cost = int(row["cost"])
            # Original balance = balance_before + cost (cost was already deducted).
            original = balance_before + cost
            delta = balance_after - original
            if delta == 0:
                continue
            out.append(
                _Event(
                    time=int(row["spin_time"]),
                    delta=delta,
                    source=SOURCE_DOUBLE_OR_NOTHING,
                    detail={
                        "outcome": "won" if bool(row["won"]) else "lost",
                        "risked": balance_before,
                    },
                )
            )
        return out

    def _tip_events(self, discord_id: int, guild_id: int | None) -> list[_Event]:
        rows = self.tip_repo.get_all_tips_for_user(discord_id, guild_id)
        out: list[_Event] = []
        for row in rows:
            amount = int(row["amount"])
            fee = int(row["fee"] or 0)
            direction = row["direction"]
            # Self-tips (sender == recipient) collapse the two sides into one row,
            # so the balance only moves by the fee.
            if row["sender_id"] == row["recipient_id"]:
                delta = -fee
            elif direction == "sent":
                delta = -(amount + fee)
            else:  # received
                delta = amount
            if delta == 0:
                continue
            out.append(
                _Event(
                    time=int(row["timestamp"]),
                    delta=delta,
                    source=SOURCE_TIPS,
                    detail={"direction": direction, "amount": amount, "fee": fee},
                )
            )
        return out

    def _disburse_events(
        self, discord_id: int, guild_id: int | None
    ) -> list[_Event]:
        rows = self.disburse_repo.get_recipient_history(discord_id, guild_id)
        return [
            _Event(
                time=int(row["disbursed_at"]),
                delta=int(row["amount"]),
                source=SOURCE_DISBURSE,
                detail={"method": row["method"]},
            )
            for row in rows
            if int(row["amount"]) != 0
        ]

    def _bonus_events(self, discord_id: int, guild_id: int | None) -> list[_Event]:
        rows = self.match_repo.get_player_bonus_events(discord_id, guild_id)
        out: list[_Event] = []
        for row in rows:
            match_time = row.get("match_time")
            if match_time is None:
                continue
            persisted = row.get("bonus_jc")
            if persisted is not None:
                # Use the actual net JC credited at match time so the chart
                # reflects garnishment / bankruptcy penalties accurately.
                delta = int(persisted)
                detail_components = {"persisted_net": delta}
            else:
                # Pre-migration rows fall back to current-config reconstruction
                # (loser gets participation, winner gets win-bonus only).
                participation = 0 if row["won"] else JOPACOIN_PER_GAME
                win = JOPACOIN_WIN_REWARD if row["won"] else 0
                delta = participation + win
                detail_components = {"participation": participation, "win": win}
            if delta == 0:
                continue
            out.append(
                _Event(
                    time=int(match_time),
                    delta=delta,
                    source=SOURCE_BONUS,
                    detail={
                        "match_id": row["match_id"],
                        "components": detail_components,
                        "won": row["won"],
                    },
                )
            )
        return out

    def _dig_events(self, discord_id: int, guild_id: int | None) -> list[_Event]:
        """Derive per-action JC deltas from the ``dig_actions`` detail JSON.

        The ``dig_actions.jc_delta`` column is honoured if populated (future-proofing
        — no code currently writes it). Otherwise we parse ``detail`` per action
        type. Known gaps (documented in the class docstring): cave-in JC losses,
        per-dig paid costs, sabotage target-side gains beyond trap-triggered returns.
        """
        if self.dig_repo is None:
            return []
        rows = self.dig_repo.get_player_jc_events(discord_id, guild_id)
        out: list[_Event] = []
        for row in rows:
            delta = _dig_row_delta(discord_id, row)
            if delta == 0:
                continue
            out.append(
                _Event(
                    time=int(row["created_at"]),
                    delta=delta,
                    source=SOURCE_DIG,
                    detail={"action_type": row["action_type"]},
                )
            )
        return out


def _dig_row_delta(discord_id: int, row: dict) -> int:
    """Best-effort extraction of a user's JC delta from one ``dig_actions`` row.

    Returns 0 when no reconstructable delta applies (unknown action type, cave-in
    dig log, target-side sabotage without trap, etc.).
    """
    is_actor = row["actor_id"] == discord_id
    is_target = row["target_id"] == discord_id

    column_delta = row.get("jc_delta") or 0
    if column_delta and is_actor:
        return int(column_delta)

    try:
        detail = json.loads(row["detail"] or "{}")
    except (json.JSONDecodeError, TypeError):
        return 0

    action = row["action_type"]

    if action == "dig" and is_actor:
        # Regular dig: ``detail.jc`` is positive earnings. Cave-in logs don't set
        # this key — the JC loss there is a separate ``add_balance`` call that
        # isn't persisted anywhere, so it's a silent gap.
        return int(detail.get("jc", 0) or 0)

    if action == "help" and is_actor:
        # Hardcoded +1 JC to the helper in ``dig_service.help_tunnel``.
        return 1

    if action == "sabotage":
        trap_triggered = bool(detail.get("trap_triggered", False))
        if is_actor:
            if trap_triggered:
                return -int(detail.get("jc_lost", 0) or 0)
            return -int(detail.get("cost", 0) or 0)
        if is_target and trap_triggered:
            # Trap-triggered sabotage: actor loses ``trap_steal = cost * 2`` and the
            # target is refunded ``cost`` (the attacker's stake). So the target's
            # gain is ``jc_lost / 2``. Non-trap sabotage doesn't move the target's
            # JC at all (only blocks), which correctly returns 0 below.
            return int(detail.get("jc_lost", 0) or 0) // 2
        return 0

    if action == "boss_fight" and is_actor:
        if detail.get("won"):
            # ``detail.jc_delta`` is the *gross* payout. For wagered wins the wager
            # is never debited elsewhere, so the credited amount is ``gross - wager``
            # (see dig_service line 3608-3614). Non-wagered wins credit the full
            # random roll. Phase-1 transitions log no ``jc_delta`` → correctly 0.
            gross = int(detail.get("jc_delta", 0) or 0)
            wager = int(detail.get("wager", 0) or 0)
            if wager > 0 and gross > 0:
                return gross - wager
            return gross
        # Loss: ``add_balance(-wager)`` is issued at loss time alongside the log,
        # so the wager is the negative delta.
        return -int(detail.get("wager", 0) or 0)

    if action == "boss_retreat" and is_actor:
        return -int(detail.get("loss", 0) or 0)

    if action == "event" and is_actor:
        # ``detail.jc_delta`` (numeric events), ``detail.jc`` (success/fail events),
        # or ``cruel_echoes`` flag (-1).
        if detail.get("cruel_echoes"):
            return -1
        if "jc_delta" in detail:
            return int(detail.get("jc_delta", 0) or 0)
        if "jc" in detail:
            return int(detail.get("jc", 0) or 0)
        return 0

    if action == "abandon" and is_actor:
        return int(detail.get("refund", 0) or 0)

    # ``prestige`` and unknown action types: no direct balance change.
    return 0

File: rust/crates/cama-app/src/balance_history_service.rs
//! Unified, chart-ready Jopacoin balance history.
//!
//! The service keeps eight independent persistence reads behind a typed port,
//! reconstructs their actual balance deltas, merges them with stable timestamp
//! ordering, and exposes [`crate::drawing::BalancePoint`] values. Python
//! remains the live repository/runtime authority during parity.

use std::collections::BTreeMap;
use std::thread;

use thiserror::Error;

use crate::drawing::BalancePoint;

pub type DiscordId = i64;
pub type GuildId = i64;

pub const SOURCE_BETS: &str = "bets";
pub const SOURCE_PREDICTIONS: &str = "predictions";
pub const SOURCE_WHEEL: &str = "wheel";
pub const SOURCE_DOUBLE_OR_NOTHING: &str = "double_or_nothing";
pub const SOURCE_TIPS: &str = "tips";
pub const SOURCE_DISBURSE: &str = "disburse";
pub const SOURCE_BONUS: &str = "bonus";
pub const SOURCE_DIG: &str = "dig";

pub const LEGACY_WIN_REWARD: i64 = 10;
pub const LEGACY_PARTICIPATION_REWARD: i64 = 5;

type EventCollector<'a> = Box<dyn Fn(DiscordId, GuildId) -> Vec<BalanceEvent> + Send + Sync + 'a>;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BetHistoryRow {
    pub bet_time: i64,
    pub profit: i64,
    pub won: bool,
    pub amount: i64,
    pub leverage: i64,
    pub match_id: i64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PredictionHistoryRow {
    pub prediction_id: i64,
    pub settle_time: Option<i64>,
    pub delta: i64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WheelHistoryRow {
    pub spin_time: i64,
    pub result: i64,
    pub outcome_metadata: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DoubleOrNothingHistoryRow {
    pub spin_time: i64,
    pub cost: i64,
    pub balance_before: i64,
    pub balance_after: i64,
    pub won: bool,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TipDirection {
    Sent,
    Received,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TipHistoryRow {
    pub timestamp: i64,
    pub amount: i64,
    pub fee: i64,
    pub direction: TipDirection,
    pub sender_id: DiscordId,
    pub recipient_id: DiscordId,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DisbursementHistoryRow {
    pub disbursed_at: i64,
    pub amount: i64,
    pub method: String,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct MatchBonusHistoryRow {
    pub match_id: i64,
    pub match_time: Option<i64>,
    pub won: bool,
    pub bonus_jc: Option<i64>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DigHistoryRow {
    pub actor_id: DiscordId,
    pub target_id: Option<DiscordId>,
    pub action_type: String,
    pub detail: Option<String>,
    pub created_at: i64,
    pub jc_delta: i64,
}

/// Raw persisted-history boundary. Concrete SQLite adapters may delegate to
/// the existing Python-compatible repositories without leaking row maps into
/// the reconstruction policy.
pub trait BalanceHistoryPort: Sync {
    fn bet_history(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BetHistoryRow>;

    fn prediction_history(
        &self,
        discord_id: DiscordId,
        guild_id: GuildId,
    ) -> Vec<PredictionHistoryRow>;

    fn wheel_history(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<WheelHistoryRow>;

    fn double_or_nothing_history(
        &self,
        discord_id: DiscordId,
        guild_id: GuildId,
    ) -> Vec<DoubleOrNothingHistoryRow>;

    fn tip_history(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<TipHistoryRow>;

    fn disbursement_history(
        &self,
        discord_id: DiscordId,
        guild_id: GuildId,
    ) -> Vec<DisbursementHistoryRow>;

    fn match_bonus_history(
        &self,
        discord_id: DiscordId,
        guild_id: GuildId,
    ) -> Vec<MatchBonusHistoryRow>;

    /// `None` means that no dig repository is installed; it is neutral.
    fn dig_history(&self, discord_id: DiscordId, guild_id: GuildId) -> Option<Vec<DigHistoryRow>>;
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum BonusComponents {
    PersistedNet(i64),
    Reconstructed { participation: i64, win: i64 },
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum EventDetail {
    Bet {
        won: bool,
        amount: i64,
        leverage: i64,
        match_id: i64,
    },
    Prediction {
        won: bool,
        prediction_id: i64,
    },
    Wheel {
        won: bool,
        vanity_tax: i64,
    },
    DoubleOrNothing {
        won: bool,
        risked: i64,
    },
    Tip {
        direction: TipDirection,
        amount: i64,
        fee: i64,
    },
    Disbursement {
        method: String,
    },
    Bonus {
        match_id: i64,
        components: BonusComponents,
        won: bool,
    },
    Dig {
        action_type: String,
    },
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BalanceEvent {
    pub time: i64,
    pub delta: i64,
    pub source: &'static str,
    pub detail: EventDetail,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BalanceSeriesPoint {
    pub event_number: i32,
    pub cumulative: i64,
    pub event: BalanceEvent,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct BalanceHistory {
    pub series: Vec<BalanceSeriesPoint>,
    pub source_totals: BTreeMap<&'static str, i64>,
}

impl BalanceHistory {
    #[must_use]
    pub fn chart_points(&self) -> Vec<BalancePoint> {
        self.series
            .iter()
            .map(|point| BalancePoint {
                event_number: point.event_number,
                cumulative: point.cumulative,
                source: point.event.source.to_owned(),
            })
            .collect()
    }

    #[must_use]
    pub fn chart_totals(&self) -> BTreeMap<String, i64> {
        self.source_totals
            .iter()
            .map(|(source, total)| ((*source).to_owned(), *total))
            .collect()
    }
}

#[derive(Debug, Error, Eq, PartialEq)]
pub enum BalanceHistoryError {
    #[error("a balance-history source worker panicked")]
    WorkerPanic,
}

#[derive(Clone, Debug)]
pub struct BalanceHistoryService<P> {
    port: P,
}

impl<P> BalanceHistoryService<P>
where
    P: BalanceHistoryPort,
{
    #[must_use]
    pub const fn new(port: P) -> Self {
        Self { port }
    }

    #[must_use]
    pub const fn port(&self) -> &P {
        &self.port
    }

    pub fn port_mut(&mut self) -> &mut P {
        &mut self.port
    }

    #[must_use]
    pub fn get_balance_event_series(
        &self,
        discord_id: DiscordId,
        guild_id: GuildId,
    ) -> BalanceHistory {
        merge_events(
            self.collectors()
                .into_iter()
                .flat_map(|collector| collector(discord_id, guild_id))
                .collect(),
        )
    }

    /// Run all eight independent source reads in parallel. Join order matches
    /// collector order, preserving synchronous stable ordering for equal
    /// timestamps.
    pub fn get_balance_event_series_concurrently(
        &self,
        discord_id: DiscordId,
        guild_id: GuildId,
    ) -> Result<BalanceHistory, BalanceHistoryError> {
        let source_events = thread::scope(|scope| {
            let handles = self
                .collectors()
                .into_iter()
                .map(|collector| scope.spawn(move || collector(discord_id, guild_id)))
                .collect::<Vec<_>>();
            handles
                .into_iter()
                .map(|handle| handle.join().map_err(|_| BalanceHistoryError::WorkerPanic))
                .collect::<Result<Vec<_>, _>>()
        })?;
        Ok(merge_events(source_events.into_iter().flatten().collect()))
    }

    fn collectors(&self) -> [EventCollector<'_>; 8] {
        [
            Box::new(|discord_id, guild_id| self.bet_events(discord_id, guild_id)),
            Box::new(|discord_id, guild_id| self.prediction_events(discord_id, guild_id)),
            Box::new(|discord_id, guild_id| self.wheel_events(discord_id, guild_id)),
            Box::new(|discord_id, guild_id| self.double_or_nothing_events(discord_id, guild_id)),
            Box::new(|discord_id, guild_id| self.tip_events(discord_id, guild_id)),
            Box::new(|discord_id, guild_id| self.disbursement_events(discord_id, guild_id)),
            Box::new(|discord_id, guild_id| self.bonus_events(discord_id, guild_id)),
            Box::new(|discord_id, guild_id| self.dig_events(discord_id, guild_id)),
        ]
    }

    fn bet_events(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BalanceEvent> {
        self.port
            .bet_history(discord_id, guild_id)
            .into_iter()
            .map(|row| BalanceEvent {
                time: row.bet_time,
                delta: row.profit,
                source: SOURCE_BETS,
                detail: EventDetail::Bet {
                    won: row.won,
                    amount: row.amount,
                    leverage: row.leverage,
                    match_id: row.match_id,
                },
            })
            .collect()
    }

    fn prediction_events(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BalanceEvent> {
        self.port
            .prediction_history(discord_id, guild_id)
            .into_iter()
            .filter(|row| row.delta != 0)
            .map(|row| BalanceEvent {
                time: row.settle_time.unwrap_or_default(),
                delta: row.delta,
                source: SOURCE_PREDICTIONS,
                detail: EventDetail::Prediction {
                    won: row.delta > 0,
                    prediction_id: row.prediction_id,
                },
            })
            .collect()
    }

    fn wheel_events(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BalanceEvent> {
        self.port
            .wheel_history(discord_id, guild_id)
            .into_iter()
            .filter_map(|row| {
                if row.result == 0 {
                    return None;
                }
                let withheld = |key: &str| {
                    row.outcome_metadata
                        .as_deref()
                        .and_then(|metadata| json_i64(metadata, key))
                        .unwrap_or_default()
                        .max(0)
                };
                let vanity_tax = withheld("vanity_tax");
                // A low-priority spinner is withheld a second, separately
                // recorded tax out of the same spin. Both must come off the
                // charted result or the wheel line over-reports the win.
                let net_result = row
                    .result
                    .saturating_sub(vanity_tax)
                    .saturating_sub(withheld("low_priority_tax"));
                Some(BalanceEvent {
                    time: row.spin_time,
                    delta: net_result,
                    source: SOURCE_WHEEL,
                    detail: EventDetail::Wheel {
                        won: net_result > 0,
                        vanity_tax,
                    },
                })
            })
            .collect()
    }

    fn double_or_nothing_events(
        &self,
        discord_id: DiscordId,
        guild_id: GuildId,
    ) -> Vec<BalanceEvent> {
        self.port
            .double_or_nothing_history(discord_id, guild_id)
            .into_iter()
            .filter_map(|row| {
                let original = row.balance_before.saturating_add(row.cost);
                let delta = row.balance_after.saturating_sub(original);
                (delta != 0).then_some(BalanceEvent {
                    time: row.spin_time,
                    delta,
                    source: SOURCE_DOUBLE_OR_NOTHING,
                    detail: EventDetail::DoubleOrNothing {
                        won: row.won,
                        risked: row.balance_before,
                    },
                })
            })
            .collect()
    }

    fn tip_events(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BalanceEvent> {
        self.port
            .tip_history(discord_id, guild_id)
            .into_iter()
            .filter_map(|row| {
                let delta = if row.sender_id == row.recipient_id {
                    -row.fee
                } else if row.direction == TipDirection::Sent {
                    -row.amount.saturating_add(row.fee)
                } else {
                    row.amount
                };
                (delta != 0).then_some(BalanceEvent {
                    time: row.timestamp,
                    delta,
                    source: SOURCE_TIPS,
                    detail: EventDetail::Tip {
                        direction: row.direction,
                        amount: row.amount,
                        fee: row.fee,
                    },
                })
            })
            .collect()
    }

    fn disbursement_events(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BalanceEvent> {
        self.port
            .disbursement_history(discord_id, guild_id)
            .into_iter()
            .filter(|row| row.amount != 0)
            .map(|row| BalanceEvent {
                time: row.disbursed_at,
                delta: row.amount,
                source: SOURCE_DISBURSE,
                detail: EventDetail::Disbursement { method: row.method },
            })
            .collect()
    }

    fn bonus_events(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BalanceEvent> {
        self.port
            .match_bonus_history(discord_id, guild_id)
            .into_iter()
            .filter_map(|row| {
                let time = row.match_time?;
                let (delta, components) = match row.bonus_jc {
                    Some(delta) => (delta, BonusComponents::PersistedNet(delta)),
                    None if row.won => (
                        LEGACY_WIN_REWARD,
                        BonusComponents::Reconstructed {
                            participation: 0,
                            win: LEGACY_WIN_REWARD,
                        },
                    ),
                    None => (
                        LEGACY_PARTICIPATION_REWARD,
                        BonusComponents::Reconstructed {
                            participation: LEGACY_PARTICIPATION_REWARD,
                            win: 0,
                        },
                    ),
                };
                (delta != 0).then_some(BalanceEvent {
                    time,
                    delta,
                    source: SOURCE_BONUS,
                    detail: EventDetail::Bonus {
                        match_id: row.match_id,
                        components,
                        won: row.won,
                    },
                })
            })
            .collect()
    }

    fn dig_events(&self, discord_id: DiscordId, guild_id: GuildId) -> Vec<BalanceEvent> {
        self.port
            .dig_history(discord_id, guild_id)
            .unwrap_or_default()
            .into_iter()
            .filter_map(|row| {
                let delta = dig_row_delta(discord_id, &row);
                (delta != 0).then_some(BalanceEvent {
                    time: row.created_at,
                    delta,
                    source: SOURCE_DIG,
                    detail: EventDetail::Dig {
                        action_type: row.action_type,
                    },
                })
            })
            .collect()
    }
}

#[must_use]
pub fn merge_events(mut events: Vec<BalanceEvent>) -> BalanceHistory {
    if events.is_empty() {
        return BalanceHistory::default();
    }
    events.sort_by_key(|event| event.time);
    let mut cumulative = 0_i64;
    let mut source_totals = BTreeMap::new();
    let series = events
        .into_iter()
        .enumerate()
        .map(|(index, event)| {
            cumulative = cumulative.saturating_add(event.delta);
            source_totals
                .entry(event.source)
                .and_modify(|total: &mut i64| *total = total.saturating_add(event.delta))
                .or_insert(event.delta);
            BalanceSeriesPoint {
                event_number: i32::try_from(index.saturating_add(1)).unwrap_or(i32::MAX),
                cumulative,
                event,
            }
        })
        .collect();
    source_totals.retain(|_, total| *total != 0);
    BalanceHistory {
        series,
        source_totals,
    }
}

#[must_use]
pub fn dig_row_delta(discord_id: DiscordId, row: &DigHistoryRow) -> i64 {
    let is_actor = row.actor_id == discord_id;
    let is_target = row.target_id == Some(discord_id);
    if row.jc_delta != 0 && is_actor {
        return row.jc_delta;
    }
    let detail = row.detail.as_deref().unwrap_or("{}");
    match row.action_type.as_str() {
        "dig" if is_actor => json_i64(detail, "jc").unwrap_or_default(),
        "help" if is_actor => 1,
        "sabotage" => {
            let trap_triggered = json_bool(detail, "trap_triggered").unwrap_or(false);
            if is_actor && trap_triggered {
                -json_i64(detail, "jc_lost").unwrap_or_default()
            } else if is_actor {
                -json_i64(detail, "cost").unwrap_or_default()
            } else if is_target && trap_triggered {
                json_i64(detail, "jc_lost").unwrap_or_default() / 2
            } else {
                0
            }
        }
        "boss_fight" if is_actor => {
            let won = json_bool(detail, "won").unwrap_or(false);
            let wager = json_i64(detail, "wager").unwrap_or_default();
            if won {
                let gross = json_i64(detail, "jc_delta").unwrap_or_default();
                if wager > 0 && gross > 0 {
                    gross.saturating_sub(wager)
                } else {
                    gross
                }
            } else {
                -wager
            }
        }
        "boss_retreat" if is_actor => -json_i64(detail, "loss").unwrap_or_default(),
        "event" if is_actor => {
            if json_bool(detail, "cruel_echoes").unwrap_or(false) {
                -1
            } else {
                json_i64(detail, "jc_delta")
                    .or_else(|| json_i64(detail, "jc"))
                    .unwrap_or_default()
            }
        }
        "abandon" if is_actor => json_i64(detail, "refund").unwrap_or_default(),
        _ => 0,
    }
}

fn json_value<'a>(raw: &'a str, key: &str) -> Option<&'a str> {
    let marker = format!("\"{key}\"");
    let start = raw.find(&marker)? + marker.len();
    let remainder = raw[start..].trim_start().strip_prefix(':')?.trim_start();
    let end = remainder.find([',', '}']).unwrap_or(remainder.len());
    Some(remainder[..end].trim())
}

fn json_i64(raw: &str, key: &str) -> Option<i64> {
    json_value(raw, key)?.parse().ok()
}

fn json_bool(raw: &str, key: &str) -> Option<bool> {
    match json_value(raw, key)? {
        "true" | "1" => Some(true),
        "false" | "0" => Some(false),
        _ => None,
    }
}

#[cfg(test)]
#[path = "balance_history_service/tests.rs"]
mod tests;
